mikrotik/resource_pool.go

```diff
@@ -5,143 +5,156 @@ import (
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	tftypes "github.com/hashicorp/terraform-plugin-framework/types"
 )
 
-func resourcePool() *schema.Resource {
-	return &schema.Resource{
-		Description: "Creates a Mikrotik IP Pool.",
+type pool struct {
+	client *client.Mikrotik
+}
 
-		CreateContext: resourcePoolCreate,
-		ReadContext:   resourcePoolRead,
-		UpdateContext: resourcePoolUpdate,
-		DeleteContext: resourcePoolDelete,
-		Importer: &schema.ResourceImporter{
-			StateContext: utils.ImportStateContextUppercaseWrapper(schema.ImportStatePassthroughContext),
-		},
+// Ensure the implementation satisfies the expected interfaces.
+var (
+	_ resource.Resource                = &pool{}
+	_ resource.ResourceWithConfigure   = &pool{}
+	_ resource.ResourceWithImportState = &pool{}
+)
+
+// NewPoolResource is a helper function to simplify the provider implementation.
+func NewPoolResource() resource.Resource {
+	return &pool{}
+}
 
-		Schema: map[string]*schema.Schema{
-			"name": {
-				Type:        schema.TypeString,
+func (r *pool) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	r.client = req.ProviderData.(*client.Mikrotik)
+}
+
+// Metadata returns the resource type name.
+func (r *pool) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_pool"
+}
+
+// Schema defines the schema for the resource.
+func (s *pool) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Creates a Mikrotik IP Pool.",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+				Description: "ID of this resource.",
+			},
+			"name": schema.StringAttribute{
 				Required:    true,
 				Description: "The name of IP pool.",
 			},
-			"ranges": {
-				Type:        schema.TypeString,
+			"ranges": schema.StringAttribute{
 				Required:    true,
 				Description: "The IP range(s) of the pool. Multiple ranges can be specified, separated by commas: `172.16.0.6-172.16.0.12,172.16.0.50-172.16.0.60`.",
 			},
-			"next_pool": {
-				Type:     schema.TypeString,
+			"next_pool": schema.StringAttribute{
 				Optional: true,
-				StateFunc: func(i interface{}) string {
-					v := i.(string)
-					// handle special case for 'none' string:
-					// it behaves the same as an empty string - unsets the value
-					// and MikroTik API will return an empty string, but we don't wont diff on '' != 'none'
-					if v == "none" {
-						return ""
-					}
-
-					return v
+				Computed: true,
+				PlanModifiers: []planmodifier.String{
```

Comment:
Is it possible to use a plan modifier to handle the `next_pool` none condition? From reading the docs, it seems like this construct may fit but I didn't get a chance to fully research it.

Reply:
@ddelnano yes, that should be possible.
I didn't extract the code to plan modifier, as it is the only case.
If you think the code would be clearer with `PlanModifier`, I can give it another shot.

Reply:
It's hard to say if it will be clearer since I don't know how it would work for this situation. However, it does seem like that would be the plugin framework's native way for doing this.

I'd at least like for us to evaluate that implementation against our current option before deciding to move forward with this.

Reply:
just checked field-level plan modifier, and it didn't work:
```
Provider "registry.terraform.io/hashicorp/mikrotik" planned an invalid value
        for mikrotik_pool.bar.next_pool: planned value cty.StringVal("none") does not
        match config value cty.StringVal("") nor prior value
        cty.StringVal("next_ip_pool-2284305059658103923").
```

Terraform expects planned value be either configured or previous which is impossible in our case.
I'll give another try with resource-level plan modifier, maybe it has different behavior 🤷

Reply:
so, I don't think it's possible at all due to data consistency rules in Terraform.
The way to go would be to update `GenericUpdateResource()` func, but I am not sure if it is a good idea with only one use case. Maybe, it we have more in the future, then I can refactor `Generic...()` functions to have more control over them from outer code.
WDYT?

Reply:
Got it. I think proceeding with what we have now makes sense. If we need this functionality in other resources then we can re-evaluate other options. Thanks for digging into this 👍